File: Python/Observer/Observer.py
#! /usr/bin/python
# -*- coding: utf-8 -*-
''' Observer Pattern
Author: Kei Nakata
Data: Oct.15.2014
'''
import abc

class Observer(object):
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def update(self):
        pass

class Subject(object):
    __metaclass__ = abc.ABCMeta

    @abc.abstractmethod
    def addObserver(self, observer):
        pass

    @abc.abstractmethod
    def notifyObserver(self):
        pass

class ConcreteObserver(Observer):
    def update(self, event):
        print self.__class__.__name__, "received from", event

class ConcreteSubject(Subject):

    def __init__(self):
        self.observers = set()

    def addObserver(self, observer):
        self.observers.add(observer)

    def notifyObserver(self):
        for observer in self.observers:
            observer.update(self.__class__.__name__)

if __name__ == '__main__':
    observer = ConcreteObserver()
    subject = ConcreteSubject()
    subject.addObserver(observer)
    subject.notifyObserver()

File: C++/Observer/Observer.cpp
/*
 * Observer Pattern
 * Author: reimen
 * Data: Oct.15.2014
 * Define a one-to-many dependency between objects so that when one
 * object changes state, all its dependents are notified and updated
 * automatically.
 */
#include <iostream>
#include <cstdlib>
#include <cstring>
#include <typeinfo>
#include <set>
using std::cout;
using std::endl;
using std::string;
using std::set;
using std::type_info;

class Observer
{
public:
  virtual ~Observer() {}
  virtual void update(const string& event) = 0;
};

class Subject
{
protected:
  set<Observer*> _observers;
public:
  virtual ~Subject() {}
  virtual void addOberver(Observer* observer) = 0;
  virtual void notifyObserver() = 0;
};

class ConcreteObserver : public Observer
{
protected:
  void update(const string& event)
  {
    const type_info& id = typeid(*this);
    cout << id.name() << " updated from " << event << endl;
  }
public:
  virtual ~ConcreteObserver() {}
};

class ConcreteSubject : public Subject
{
public:
  virtual ~ConcreteSubject() {}
  virtual void addOberver(Observer* observer)
  {
    _observers.insert(observer);
  }
  virtual void notifyObserver()
  {
    for(set<Observer*>::iterator it = _observers.begin(); it != _observers.end(); ++it)
      {
        const type_info& id = typeid(*this);
        string event = id.name();
        (*it)->update(event);
      }
  }
};


int main()
{
  ConcreteSubject subject;
  ConcreteObserver observer_a, observer_b, observer_c;
  subject.addOberver(&observer_a);
  subject.addOberver(&observer_b);
  subject.addOberver(&observer_c);
  subject.notifyObserver();
  return EXIT_SUCCESS;
}
